src/dimmwitted.cc

```diff
@@ -184,36 +295,52 @@ void DimmWitted::learn() {
     for (auto &sampler : samplers) sampler.wait();
 
     // sum the weights and store in the first factor graph
-    // the average weights will be calculated and assigned to all factor graphs
+    // the average weights will be calculated
     for (size_t i = 1; i < n_samplers_; ++i)
       infrs.merge_weights_from(samplers[i].infrs);
     if (n_samplers_ > 1) {
       infrs.average_weights(n_samplers_);
     }
-    for (size_t i = 1; i < n_samplers_; ++i)
-      infrs.copy_weights_to(samplers[i].infrs);
 
     // calculate the norms of the difference of weights from the current epoch
     // and last epoch
     double lmax = -INFINITY;
     double l2 = 0.0;
     for (size_t j = 0; j < nweight; ++j) {
-      double diff = fabs(prev_weights[j] - infrs.weight_values[j]);
-      prev_weights[j] = infrs.weight_values[j];
+      double delta_j = infrs.weight_values[j] - prev_weights[j];
+      delta[j] = delta_j;
+      double diff = fabs(delta_j);
       l2 += diff * diff;
       if (lmax < diff) lmax = diff;
     }
     lmax /= current_stepsize;
 
     double elapsed = t.elapsed();
     if (should_show_progress) {
-      std::cout << "" << elapsed << " sec.";
-      std::cout << "," << (nvar * n_samplers_) / elapsed << " vars/sec."
+      std::streamsize ss = std::cout.precision();
+      std::cout << std::setprecision(3) << "" << elapsed << " sec."
+                << "," << (nvar * n_samplers_) / elapsed << " vars/sec."
                 << ",stepsize=" << current_stepsize << ",lmax=" << lmax
-                << ",l2=" << sqrt(l2) / current_stepsize << std::endl;
+                << ",l2=" << sqrt(l2) / current_stepsize << std::endl
+                << std::setprecision(ss);
     }
 
+    // in distributed mode: exchange local gradients for latest weights
+    bool stop = false;
+    if (is_distributed()) {
+      stop = ps_update_weights(n_samplers_ * (i_epoch + 1), delta);
```

Comment:
This may be more of a style issue, but..  I wish we could make this for-loop body appear mostly straight-line code.  Folding most of the `is_distributed()` conditional branches into subroutines, namely `ps_update_weights()`, to do early returns as you did for `ps_send_message()` will do the job.  The `stop` flag could stay outside of the loop and be the loop condition that always stays false if not distributed, which will also remove the `break` below.